dds/DCPS/transport/rtps_udp/RtpsUdpReceiveStrategy.cpp

```diff
@@ -132,9 +151,135 @@ RtpsUdpReceiveStrategy::receive_bytes(iovec iov[],
 #endif
   remote_address_ = remote_address;
 
+#ifdef OPENDDS_SECURITY
+  if (stop) {
+    return ret;
+  }
+
+  using namespace DDS::Security;
+  const ParticipantCryptoHandle receiver = link_->local_crypto_handle();
+  if (ret > 0 && receiver != DDS::HANDLE_NIL) {
+    encoded_rtps_ = false;
+
+    const CryptoTransform_var crypto = link_->security_config()->get_crypto_transform();
+    if (!crypto) {
+      return recv_err("no crypto plugin", remote_address, stop);
+    }
+
+    if (ret < RTPS::RTPSHDR_SZ + RTPS::SMHDR_SZ) {
+      return recv_err("message too short", remote_address, stop);
+    }
+
+    const unsigned int encLen = static_cast<unsigned int>(ret);
+    DDS::OctetSeq encoded(encLen);
+    encoded.length(encLen);
+    unsigned char* const encBuf = encoded.get_buffer();
+    size_t copied = 0;
+    for (int i = 0; i < n && copied < encLen; ++i) {
+      const size_t chunk = std::min(static_cast<size_t>(iov[i].iov_len),
+                                    static_cast<size_t>(encLen - copied));
+      std::memcpy(encBuf + copied, iov[i].iov_base, chunk);
+      copied += chunk;
+    }
+
+    if (copied != encLen) {
+      return recv_err("received bytes didn't fit in iovec array", remote_address, stop);
+    }
+
+    if (encoded[RTPS::RTPSHDR_SZ] != RTPS::SRTPS_PREFIX) {
+      return ret;
+    }
+
+    GUID_t peer;
+    static const int GuidPrefixOffset = 8; // "RTPS", Version(2), Vendor(2)
+    std::memcpy(peer.guidPrefix, encBuf + GuidPrefixOffset, sizeof peer.guidPrefix);
+    peer.entityId = RTPS::ENTITYID_PARTICIPANT;
+    const ParticipantCryptoHandle sender = link_->peer_crypto_handle(peer);
+    if (sender == DDS::HANDLE_NIL) {
+      if (security_debug.warn) {
+        ACE_DEBUG((LM_WARNING, "decode_rtps_message no remote participant crypto handle, dropping\n"));
+      }
+      stop = true;
+      return ret;
+    }
+
+    DDS::OctetSeq plain;
+    SecurityException ex = {"", 0, 0};
+    if (!crypto->decode_rtps_message(plain, encoded, receiver, sender, ex)) {
+      if (security_debug.warn) {
+        ACE_DEBUG((LM_WARNING, "decode_rtps_message SecurityException [%d.%d]: %C\n",
+                   ex.code, ex.minor_code, ex.message.in()));
+      }
+      if (ex.code == -10 && ex.minor_code == 1) {
```

Comment:
If we're checking these security exception codes, maybe they need to be enums / constants for clarity?

Reply:
I agree it's odd but I'm not sure what the best way to fix it is.  In theory, another plugin could be in use here, and the spec is silent about error codes.

Because key exchange happens asynchronously, we don't know if the key is available at any given time.  If we don't have the key yet, we drop the message without logging an error.  The plugin implementation doesn't have a way to (standardly) report this state.